### ISSUE TITLE
Any luck finetuning?

### ISSUE THREAD

**Issue by ardywibowo** [Lightning-AI/lit-llama#131]
I'm currently running `finetune_lora.py` to test reproducibility on the Alpaca dataset. Currently at around iteration 13,000 and it has not output reasonable responses to the validation prompt yet. How long do you typically finetune until you start seeing generations that make sense?

Some details of my setup:
- I'm using an older GPU that doesn't support bfloat16.
- I changed the setup to use 32-bit precision and changed the training strategy to DeepSpeed (~6x slowdown).
- Disabled gradient clipping since DeepSpeed handles that internally (not sure how to change the DeepSpeed config to enabled that though, anyone have any ideas?)

Thanks!

**Comment by lantiga** [comment]
Take a look at this improved branch for LLaMA-Adapter: https://github.com/Lightning-AI/lit-llama/pull/128, this produces very high-quality generations and it converges quickly. We'll be merging this in the next few hours.

We'll be looking at optimizing LoRA next.

**Comment by ardywibowo** [comment]
Thanks! I will check it out. 

Any tips on using `16-mixed` vs. `bfloat16` (my GPU doesn't support `bfloat16`)? I think it shouldn't make much of a difference, but any anecdotal data points of figuring out how to set hyperparameters to finetune this would be very useful :D

**Comment by lantiga** [comment]
#128 has now landed, you should be able to just change `dtype` to `float16` in the `finetune_adapter.py` script without major issues I believe.

**Comment by SHENZHENYI** [comment]
I have run 'finetune_lora.py' for 7,5000 iterations when the micro-batch size is 2. (lora-r=16)

Don't know if your issue comes from #125.

And when you do generation. The current generate function doesn't stop when it hits <eos>. I fixed it myself.

Below are the prompted results. I think they are fairly good.

'''
### Instruction:

Tell me about alpacas.

### Response:

Alpacas are a type of domesticated animal that originated in the high altitude areas of South America. They belong to the camelid family, and are related to camels, llamas, and vicuñas. They are relatively small and stocky, with shaggy, long coats. Their coat comes in two varieties: suri and huacaya. Suris have long, silky, and curly coats, while huacayas have coarser, flat, and dense coats. Alpacas are bred for their meat, wool, and hide, and are raised for their fiber. They are known for their docile temperament and are often considered domesticated animals. They are typically kept on small farms or ranches in their natural habitat.

### Instruction:

Tell me about the president of Mexico in 2019.

### Response:

Most of the information you may wish to know about President Andrés Manuel López Obrador, commonly known as Andrés Manuel or AMLO, is already available in many online sources.

A brief summary of the current president of Mexico is as follows:

- Educated as lawyer and public security specialist
- Political career: he was a Member of the Mexican Congress between 2000 and 2005.
- He served as the Mayor of Mexico City from 2000 to 2005.
- Elected president of Mexico in the general election of 2018.
- Assuming the presidential term on 1 December 2018, with a majority in both the Chamber of Deputies and the Senate.
- In 2020 was recognized by the United Nations as a Champion of the Sustainable Development Goals for his work towards theimplementation of the SDGs agenda.
- Successively, he reached an agreement with the business community to share the blame for the economic catastrophe of theearly 2000s and discharge the credit and debt for its clients in 2005.

### Instruction:

Tell me about the king of France in 2019.

### Response:

The current king of France is Emmanuel Macron. He was elected in 2017 and was sworn in as the President of France on May 14, 2017. King Macron has not been on the throne since 1793, when the French monarchy was abolished during the time of the French Revolution. Since then, France has had the office of President. The most recent King of France was Charles X.

### Instruction:

List all Canadian provinces in alphabetical order.

### Response:

The list of Canadian Provinces in alphabetical order is:

1. Alberta
2. British Columbia
3. New Brunswick
4. Newfoundland and Labrador
5. Nova Scotia
6. Ontario
7. Prince Edward Island
8. Quebec
9. Saskatchewan
10. Yukon.
'''

**Comment by ardywibowo** [comment]
Thanks! I ran the new code and it seems to be stable.

For others who don’t have a bfloat16 GPU, I ran ‘finetune_adapter.py’ with some parameters float and some params half-precision, and needed to use Deepspeed Stage 3 to get it to fit on my GPU.
